File: frida_trainer/player.py
import scan


class Player:
    def __init__(self, ptr, mem_helper):
        self._ptr = ptr
        self._offsets = {
             'health': [0xf8],
             'weapon_ammo': [0x374, 0x10, 0x0],
             'weapon_clip': [0x374, 0x14, 0x0],
             'player_name': [0x225],
             'head_vector3f': [0x4],
             'foot_vector3f': [0x34],
             'yaw': [0x40],
             'pitch': [0x44]
         }
        self._m = mem_helper
        self.name = self._m.read_mem('s', self._m.resolve_ptr_list(self._ptr, self._offsets['player_name']))

    @property
    def health(self):
        health_addr = self._m.resolve_ptr_list(self._ptr, self._offsets['health'])
        if health_addr < 0x1000:
             return -1, -1
        return health_addr, self._m.read_mem('s32', health_addr)

    @health.setter
    def health(self, new_health):
        health_addr, cur_health = self.health
        print('[*] Current health: {}'.format(cur_health))
        print('[+] Overwriting with: {}'.format(new_health))
        self._m.write_mem('u32', health_addr, new_health)
        _, cur_health = self.health
        print('[*] New health: {}'.format(cur_health))

    @property
    def clip(self):
        clip_addr = self._m.resolve_ptr_list(self._ptr, self._offsets['weapon_clip'])
        return clip_addr, self._m.read_mem('u32', clip_addr)

    @clip.setter
    def clip(self, new_clip):
        clip_addr, cur_clip = self.clip
        print('[*] Current clip: {}'.format(cur_clip))
        print('[+] Overwriting with: {}'.format(new_clip))
        self._m.write_mem('u32', clip_addr, new_clip)
        _, cur_clip = self.clip
        print('[*] New clip: {}'.format(cur_clip))

    @property
    def ammo(self):
        ammo_addr = self._m.resolve_ptr_list(self._ptr, self._offsets['weapon_ammo'])
        return ammo_addr, self._m.read_mem('u32', ammo_addr)

    @ammo.setter
    def ammo(self, new_ammo):
        ammo_addr, cur_ammo = self.ammo
        print('[*] Current ammo: {}'.format(cur_ammo))
        print('[+] Overwriting with: {}'.format(new_ammo))
        self._m.write_mem('u32', ammo_addr, new_ammo)
        _, cur_ammo = self.ammo
        print('[*] New ammo: {}'.format(cur_ammo))

    @property
    def pitch(self):
        pitch_addr = self._m.resolve_ptr_list(self._ptr, self._offsets['pitch'])
        return pitch_addr, self._m.read_mem('f', pitch_addr)

    @pitch.setter
    def pitch(self, new_pitch):
        pitch_addr, cur_pitch = self.pitch
        print('[*] Current pitch: {}'.format(cur_pitch))
        print('[+] Overwriting with: {}'.format(new_pitch))
        self._m.write_mem('f', pitch_addr, new_pitch)
        _, cur_pitch = self.pitch
        print('[*] New pitch: {}'.format(cur_pitch))

    @property
    def yaw(self):
        yaw_addr = self._m.resolve_ptr_list(self._ptr, self._offsets['yaw'])
        return yaw_addr, self._m.read_mem('f', yaw_addr)

    @yaw.setter
    def yaw(self, new_yaw):
        yaw_addr, cur_yaw = self.yaw
        print('[*] Current yaw: {}'.format(cur_yaw))
        print('[+] Overwriting with: {}'.format(new_yaw))
        self._m.write_mem('f', yaw_addr, new_yaw)
        _, cur_yaw = self.yaw
        print('[*] New yaw: {}'.format(cur_yaw))

    def get_position(self, location='head'):
        assert location in ['head', 'foot']
        position_ptr = self._m.resolve_ptr_list(self._ptr, self._offsets['{}_vector3f'.format(location)])
        x = self._m.read_mem('f', position_ptr)
        y = self._m.read_mem('f', position_ptr + scan.mem_types['f'][1])
        z = self._m.read_mem('f', position_ptr + scan.mem_types['f'][1] * 2)
        return x, y, z

    def is_alive(self):
        _, player_health = self.health
        return (player_health > 0) and (player_health <= 100)

File: native_trainer/Headshot/player.cpp
//
//  player.cpp
//  Headshot
//
//  Created by Jai  Verma on 20/10/18.
//  Copyright © 2018 Jai  Verma. All rights reserved.
//

#include "player.hpp"

Player::Player(uintptr_t addr) {
    this->addr = addr;
    constants = Constants();
}

unsigned int Player::get_health(task_t task) {
    auto health_addr = resolve_pointer_chain(task, addr, constants.offsets["health"]);
    if (health_addr < 0x1000)
        return 101;
    return read_data<uint32_t>(task, health_addr);
}

void Player::set_health(task_t task, unsigned int health) {
    auto health_addr = resolve_pointer_chain(task, addr, constants.offsets["health"]);
    auto new_health = static_cast<uint32_t>(health);
    write_data(task, health_addr, new_health);
}

unsigned int Player::get_clip(task_t task) {
    auto clip_addr = resolve_pointer_chain(task, addr, constants.offsets["weapon_clip"]);
    return read_data<uint32_t>(task, clip_addr);
}

void Player::set_clip(task_t task, unsigned int ammo) {
    auto clip_addr = resolve_pointer_chain(task, addr, constants.offsets["weapon_clip"]);
    auto new_clip = static_cast<uint32_t>(ammo);
    write_data(task, clip_addr, new_clip);
}

unsigned int Player::get_ammo(task_t task) {
    auto ammo_addr = resolve_pointer_chain(task, addr, constants.offsets["weapon_ammo"]);
    return read_data<uint32_t>(task, ammo_addr);
}

void Player::set_ammo(task_t task, unsigned int ammo) {
    auto ammo_addr = resolve_pointer_chain(task, addr, constants.offsets["weapon_ammo"]);
    auto new_ammo = static_cast<uint32_t>(ammo);
    write_data(task, ammo_addr, new_ammo);
}

float Player::get_pitch(task_t task) {
    auto pitch_addr = resolve_pointer_chain(task, addr, constants.offsets["pitch"]);
    return read_data<float>(task, pitch_addr);
}

void Player::set_pitch(task_t task, float pitch) {
    auto pitch_addr = resolve_pointer_chain(task, addr, constants.offsets["pitch"]);
    write_data(task, pitch_addr, pitch);
}

float Player::get_yaw(task_t task) {
    auto yaw_addr = resolve_pointer_chain(task, addr, constants.offsets["yaw"]);
    return read_data<float>(task, yaw_addr);
}

void Player::set_yaw(task_t task, float yaw) {
    auto yaw_addr = resolve_pointer_chain(task, addr, constants.offsets["yaw"]);
    write_data(task, yaw_addr, yaw);
}

bool Player::is_alive(task_t task) {
    unsigned int health = get_health(task);
    return (health <= 100);
}

std::tuple<float, float, float> Player::get_position(task_t task, std::string location) {
    uintptr_t position_addr = NULL;
    if (location == "head")
        position_addr = resolve_pointer_chain(task, addr, constants.offsets["head_vector3f"]);
    else // assume position to be foot
        position_addr = resolve_pointer_chain(task, addr, constants.offsets["foot_vector3f"]);
    
    auto x = read_data<float>(task, position_addr);
    auto y = read_data<float>(task, position_addr + sizeof(float));
    auto z = read_data<float>(task, position_addr + sizeof(float) * 2);
    
    return std::make_tuple(x, y, z);
}
